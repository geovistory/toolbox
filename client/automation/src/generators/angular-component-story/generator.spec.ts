import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import angularComponentStoryGenerator from './generator';

describe('angular-component-story generator', () => {
  let tree: Tree;
  const libName = 'ng-lib1';
  const storyFile = `${libName}/src/lib/test-button/test-button.component.stories.ts`;
  const componentFolder = `${libName}/src/lib/test-button`
  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    tree.write(
      `${libName}/src/lib/test-button/test-button.component.ts`,
      `import { Component, Input } from '@angular/core';

        export type ButtonStyle = 'default' | 'primary' | 'accent';

        @Component({
          selector: 'proj-test-button',
          templateUrl: './test-button.component.html',
          styleUrls: ['./test-button.component.css']
        })
        export class TestButtonComponent {
          @Input('buttonType') type = 'button';
          @Input() style: ButtonStyle = 'default';
          @Input() age?: number;
          @Input() isOn = false;
          @Input() message: string | undefined;
          @Input() anotherProp: any;
          @Input() anotherNeverProp: never;
        }`
    );
  });

  it('should not generate the component stories file when it already exists', async () => {
    tree.write(storyFile, '');
    await angularComponentStoryGenerator(tree, { componentPath: componentFolder });
    expect(tree.read(storyFile, 'utf-8')).toBe('');
  });

  it('should generate the component stories file', async () => {
    await angularComponentStoryGenerator(tree, { componentPath: componentFolder });
    expect(tree.exists(storyFile)).toBe(true);
  });

  it('should generate the right stories file', async () => {
    await angularComponentStoryGenerator(tree, { componentPath: componentFolder });
    expect(tree.read(storyFile)?.toString()).toMatchSnapshot();
  });
});
