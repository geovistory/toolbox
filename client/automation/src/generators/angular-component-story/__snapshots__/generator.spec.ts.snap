// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`angular-component-story generator should generate the right stories file 1`] = `
"import {
  applicationConfig,
  type Meta,
  type StoryObj,
} from '@storybook/angular';
import { TestButtonComponent } from './test-button.component';
import { importProvidersFrom } from '@angular/core';
import { MatDialogModule } from '@angular/material/dialog';
import { provideAnimations } from '@angular/platform-browser/animations';
import { StateModule } from '@kleiolab/lib-redux';
import { INITIAL_STATE } from '@ngrx/store';

import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

import { MockStateFactory } from './../../.storybook/MockStateFactory';
import { ActiveProjectService } from './services/active-project.service';

const meta: Meta<TestButtonComponent> = {
  component: TestButtonComponent,
  title: 'TestButtonComponent',
  decorators: [
    applicationConfig({
      providers: [
        provideAnimations(),
        importProvidersFrom(StateModule, MatDialogModule),
        ActiveProjectService,
      ],
    }),
  ],
};
export default meta;
type Story = StoryObj<TestButtonComponent>;

// Setup the initial state of the story Basic
const stateBasic = new MockStateFactory();

export const Basic: Story = {
  args: {
    buttonType: 'button',
    style: 'default',
    age: 0,
    isOn: false,
  },
  decorators: [
    applicationConfig({
      providers: [{ provide: INITIAL_STATE, useValue: stateBasic.state }],
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(/test-button works!/gi)).toBeTruthy();
  },
};
"
`;
